timeboost/s3_storage.go

```diff
@@ -0,0 +1,218 @@
+			log.Error("error deleting s3-persisted bids from sql db using lastFailedDeleteRound", "lastFailedDeleteRound", s.lastFailedDeleteRound, "err", err)
+			return 5 * time.Second
+		}
+		s.lastFailedDeleteRound = 0
+	}
+
+	bids, round, err := s.sqlDB.GetBids(s.config.MaxDbRows)
+	if err != nil {
+		log.Error("Error fetching validated bids from sql DB", "round", round, "err", err)
+		return 5 * time.Second
+	}
+	// Nothing to persist or a contiguous set of bids wasn't found, so exit early
+	if len(bids) == 0 {
+		return s.config.UploadInterval
+	}
+
+	var csvBuffer bytes.Buffer
+	var size int
+	var firstBidId int
+	csvWriter := csv.NewWriter(&csvBuffer)
+	uploadAndDeleteBids := func(firstRound, deletRound uint64) error {
+		// End current batch when size exceeds MaxBatchSize and the current round ends
+		csvWriter.Flush()
+		if err := csvWriter.Error(); err != nil {
+			log.Error("Error flushing csv writer", "err", err)
+			return err
+		}
+		if err := s.uploadBatch(ctx, csvBuffer.Bytes(), firstRound); err != nil {
+			log.Error("Error uploading batch to s3", "firstRound", firstRound, "err", err)
+			return err
+		}
+		// After successful upload we should go ahead and delete the uploaded bids from DB to prevent duplicate uploads
+		// If the delete fails, we track the deleteRound until a future delete succeeds.
+		if err := s.sqlDB.DeleteBids(deletRound); err != nil {
+			log.Error("error deleting s3-persisted bids from sql db", "round", deletRound, "err", err)
+			s.lastFailedDeleteRound = deletRound
+		} else {
+			// Previously failed deletes dont matter anymore as the recent one (larger round number) succeeded
+			s.lastFailedDeleteRound = 0
+		}
+		return nil
+	}
+
+	header := []string{"ChainID", "Bidder", "ExpressLaneController", "AuctionContractAddress", "Round", "Amount", "Signature"}
+	if err := csvWriter.Write(header); err != nil {
+		log.Error("Error writing to csv writer", "err", err)
+		return 5 * time.Second
+	}
+	for index, bid := range bids {
+		record := []string{bid.ChainId, bid.Bidder, bid.ExpressLaneController, bid.AuctionContractAddress, fmt.Sprintf("%d", bid.Round), bid.Amount, bid.Signature}
+		if err := csvWriter.Write(record); err != nil {
+			log.Error("Error writing to csv writer", "err", err)
+			return 5 * time.Second
+		}
+		if s.config.MaxBatchSize != 0 {
+			size += csvRecordSize(record)
+			if size >= s.config.MaxBatchSize && index < len(bids)-1 && bid.Round != bids[index+1].Round {
+				if uploadAndDeleteBids(bids[firstBidId].Round, bids[index+1].Round) != nil {
+					return 5 * time.Second
+				}
+				// Reset csv for next batch
+				csvBuffer.Reset()
+				if err := csvWriter.Write(header); err != nil {
+					log.Error("Error writing to csv writer", "err", err)
+					return 5 * time.Second
+				}
+				size = 0
+				firstBidId = index + 1
+			}
+		}
+	}
+	if s.config.MaxBatchSize == 0 || size > 0 {
+		if uploadAndDeleteBids(bids[firstBidId].Round, round) != nil {
```

Comment:
I was concerned about a potential race condition in uploadBatches between getting maxRound and deleting bids WHERE Round < maxRound, but thinking it through: bids can only come in for the current active round, and since we query for Round < maxRound, we're never looking at the active round's bids. So there shouldn't actually be a way for new bids to appear for the rounds we're processing and deleting, right?

Reply:
Yep, there shouldn't be a race here for current round's bids